src/Tutor.Core/Domain/KnowledgeMastery/SessionTracker.cs

```diff
@@ -14,28 +14,59 @@ public class SessionTracker : EventSourcedEntity
     public TimeSpan DurationOfFinishedSessions { get; private set; } = new(0);
     public bool HasUnfinishedSession => StartOfUnfinishedSession.HasValue;
 
-    public TimeSpan? DurationOfUnfinishedSession => HasUnfinishedSession ? LastActivity.Value - StartOfUnfinishedSession.Value : null;
+    public TimeSpan? DurationOfUnfinishedSession => HasUnfinishedSession ? DateTime.UtcNow - StartOfUnfinishedSession.Value : null;
     public DateTime? StartOfUnfinishedSession { get; private set; }
-    public DateTime? LastActivity { get; private set; }
+
+    public bool IsPaused { get; private set; }
+    public TimeSpan DurationOfAllPauses => DurationOfFinishedPauses + DurationOfUnfinishedPauses.GetValueOrDefault();
+    public TimeSpan DurationOfFinishedPauses { get; private set; } = new(0);
+    public TimeSpan? DurationOfUnfinishedPauses => IsPaused ? DateTime.UtcNow - LastPause.Value : null;
+    public DateTime? LastPause { get; private set; }
 
     public void Launch()
     {
-        if (HasUnfinishedSession) Causes(new SessionAbandoned());
-
+        if (HasUnfinishedSession)
+        {
+            if (IsPaused) Continue();
+            Causes(new SessionAbandoned());
+        }
+        
         Causes(new SessionLaunched());
     }
 
     public Result Terminate()
     {
         if (!HasUnfinishedSession) return Result.Fail("No active session to terminate.");
-
+        if (IsPaused) Continue();
+        
         Causes(new SessionTerminated());
         return Result.Ok();
     }
 
+    public Result Pause()
+    {
+        if (!HasUnfinishedSession) return Result.Fail("No active session to pause.");
+        if (IsPaused) return Result.Fail("Session is already paused.");
+        
+        IsPaused = true;
+        Causes(new SessionPaused());
+        return Result.Ok();
+    }
+
+    public Result Continue()
+    {
+        if (!HasUnfinishedSession) return Result.Fail("No active session to continue.");
+        if (!IsPaused) return Result.Fail("Session is not paused.");
```

Comment:
Question: Should this check be placed inside Continue method or inside methods that call it? Currently, both places have the same checks.
Opinion: The check is related to the Continue logic, so it should be here. To avoid code duplication, we could do the following: Methods that call Continue don't check if the session is paused, just call the Continue method and return if it is failed.
Current state: Method that call Continue check if IsPaused and only call if correct. Then Continue also checks if  IsPaused.

We still have an if in the method that calls Continue, but we don't double the same check.
Not sure if this is the right approach, @Luburic should comment.

Reply:
The final check should be in the `Continue` method to ensure a valid state.

However, the `IsPaused` check in, for example, the `Terminate` method above (line 40) is not a validation rule but a control flow rule. We could remove the conditional expression and just call Continue, knowing it will fail when it needs to, but such code is less expressive.